Implement Pager for auto-paging requests

t8y8 opened the pull request:
Here's a first draft based on @RussTheAerialist's code in `pagination_sample.py`.

Currently it:
- Takes an endpoint as its argument instead of a `get` function, this is because these calls only apply to `get` anyway, so let's keep it clean.
- Takes an options object and seeds the filters/sorts with it, as well as the initial page size
- Can be used as an iterator

I would like it to:

- Not do everything in the `__call__` method. I _think_ we can move that to a helper or to `__init__`
    - We would make the first page call on instantiation, which populates some properties that could be useful (like `__len__`)
- Since we know the `total_avaliable` we might be able to implement `__len__`. This gives users the ability to check the length to see if they really want to iterate over it or not.
- I need to read more to see if I need to worry about `__next__` or not.
- Add tests

This script works (tested manually):
```python
import tableauserverclient as TSC

auth_stuff = TSC.TableauAuth(username='USERNAME', password='PASSWORD')

server = TSC.Server("http://SERVER")
server.auth.sign_in(auth_stuff)

options = TSC.RequestOptions()
options.filter.add(TSC.Filter(TSC.RequestOptions.Field.Name,
                                         TSC.RequestOptions.Operator.Equals,
                                         "WORKBOOKNAME"))

# Call with a filter
wbs = TSC.server.Pager(server.workbooks, options)
print(len([i for i in wbs[)) # prints 1

# Get them all!
wbs = TSC.server.Pager(server.workbooks)
print(len([i for i in wbs])) # prints 1086 on my test server, that's all of em!
server.auth.sign_out()
```

Open Questions:

1. Where should it live? I'm not sure Server is the right place
2. Thoughts on implementing `__len__`?
3. Thoughts on implementing the primary logic somewhere other than `__call__`? It might just be me who thinks it's weird.

/cc @RussTheAerialist @LGraber

RussTheAerialist commented on tableauserverclient/server/server.py:
```diff
@@ -1,8 +1,47 @@
+    """
+
+    def __init__(self, fetcher, opts=None):
+        self._fetcher = fetcher.get
```
probably name the argument that is passed in `endpoint`

RussTheAerialist commented on tableauserverclient/server/server.py:
```diff
@@ -1,8 +1,47 @@
+        self._fetcher = fetcher.get
+        self._options = opts
+
+    def __call__(self):
```
yeah this is not ideal.  It should probably just live in `__iter__` and we get rid of `__call__`

RussTheAerialist commented on tableauserverclient/server/server.py:
```diff
@@ -1,8 +1,47 @@
+        self._options = opts
+
+    def __call__(self):
+        current_item_list, last_pagination_item = self._fetcher(self._options)
```
this won't work if they set a page number in the options because count will never be greater than total available until we hit the end and throw an error.

We should check that a page number hasn't been set in the options, and if it has output a warning or throw an exception.

t8y8 replied:
There's always a page number (it's defaulted to 1).

Should the logic be:

1. On first iteration, run the generator
2. Store some state that says "we're working on it now"
3. Then just exhaust the generator starting from that point, then they could, if they want, start on page 3

LGraber commented on tableauserverclient/server/server.py:
```diff
@@ -1,8 +1,47 @@
+class Pager(object):
+    """ This class returns a generator that will iterate over all of the results.
+
+    server is the server object that will be used when calling the callback.  It will be passed
```
This comment does not appear to be relevant anymore

LGraber replied:
why doesn't it work if we simply change how we initialize count?

LGraber commented on tableauserverclient/server/server.py:
```diff
@@ -1,8 +1,47 @@
+            if len(current_item_list) == 0:
+                current_item_list, last_pagination_item = self._load_next_page(current_item_list, last_pagination_item)
+
+            yield current_item_list.pop(0)
```
This needs a bit more protection. Technically _load_next_page can return an empty list ... even after we fix our counting issue. This could happen if someone deletes an item while we are iterating. We should make sure the list is not empty before popping.

LGraber commented:
Oh yeah ... why did you put this in the server.py file and not its own file?

t8y8 commented:
Alright! Addressed feedback (in person, and on here) tests added, and tested manually with this code:
```python
import tableauserverclient as TSC


auth_stuff = TSC.TableauAuth(username='USER', password='password')

server = TSC.Server("http://server")

server.auth.sign_in(auth_stuff)

print(server.auth_token)

options = TSC.RequestOptions(pagenumber=2, pagesize=300)

wbs = TSC.Pager(server.workbooks, options)
print("GOT WBS")
print(len({i.id for i in wbs}))
server.auth.sign_out()
```

LGraber commented on test/test_pager.py:
```diff
@@ -0,0 +1,90 @@
+            self.assertTrue(len(list(workbooks)) == 3)
+
+            # Let's check that workbook items aren't duplicates
+            workbooks = TSC.Pager(self.server.workbooks)
```
why did you need to call this again?

RussTheAerialist replied:
generators are one time use.  Once you iterate over them, you cannot go back.  He could have stored the output of `list(workbooks)` and then he wouldn't have had to call it twice.

RussTheAerialist replied:
though, it might be better (and more instructive) to use

```python
self.assertEquals(3, itertools.count(workbooks)
```

LGraber commented on test/test_pager.py:
```diff
@@ -0,0 +1,90 @@
+            # Register Pager with some pages
+            m.get(self.baseurl + "?pageNumber=1&pageSize=1", text=page_1)
+            m.get(self.baseurl + "?pageNumber=2&pageSize=1", text=page_2)
+            m.get(self.baseurl + "?pageNumber=3&pageSize=1", text=page_3)
```
why are you registering these if we are not supposed to call them?

LGraber replied:
Hahahaha. :) I know what it does. My comment was that we don't call these urls in this test so why do we need to register them. :)

LGraber replied:
My bad ... it does get called (except maybe the first one). OOps

LGraber commented on test/test_pager.py:
```diff
@@ -0,0 +1,90 @@
+            page_3 = f.read().decode('utf-8')
+        with requests_mock.mock() as m:
+            # Register Pager with default request options
+            m.get(self.baseurl, text=page_1)
```
Why are you registering this if the test is not supposed to call it?

LGraber commented:
I am confused because the tests don't represent how we expect the users to use this class (ie you keep casting to list and doing len) and you don't have a sample of how to use it. Can you please add a sample

RussTheAerialist commented on test/test_pager.py:
```diff
@@ -37,11 +37,10 @@ def test_pager_with_no_options(self):
-            workbooks = TSC.Pager(self.server.workbooks)
-            self.assertTrue(len(list(workbooks)) == 3)
+            workbooks = list(TSC.Pager(self.server.workbooks))
+            self.assertTrue(len(workbooks) == 3)
```
nitpick
```python
self.assertEquals(len(workbooks), 3)
```
